zookeeper/component.py

```diff
@@ -0,0 +1,314 @@
+        for k, v in component_annotations:
+            # The value from the `conf` dict takes priority.
+            if k in conf:
+                instance = conf["k"]
+                # Check that the configuration instance has the correct type.
+                if not issubclass(instance, v):
+                    raise ValueError(
+                        f"The configured value '{instance}' for annotated parameter '{self.__component_name__}.{k}' must be an instance of '{v.__name__}'."
+                    )
+                setattr(self, k, instance)
+
+            # If there's no config value but the value is set on the object, add
+            # that value to the config so that it can get picked up in child
+            # components.
+            elif hasattr(self, k):
+                conf[k] = getattr(self, k)
+
+            # If there's only one concrete subclass of `v`, instantiate an
+            # instance of that class.
+            elif len(get_concrete_subclasses(v)) == 1:
+                component_cls = list(get_concrete_subclasses(v))[0]
+                print_formatted_text(
+                    f"{component_cls} is the only concrete component class that satisfies the type of the annotated parameter '{self.__component_name__}.{k}'. Using an instance of this class by default."
+                )
+                # This is safe because we ban overriding `__init__`.
+                instance = component_cls()
+                setattr(self, k, instance)
+                conf[k] = instance
+
+            # If we are running interactively and there's more than one concrete
+            # subclass of `v`, prompt for the concrete subclass to instantiate.
+            # Add the instance to the configuation so that is can get passed to
+            # any children.
+            elif interactive and len(get_concrete_subclasses(v)) > 1:
+                component_cls = prompt_for_component(
+                    f"{self.__component_name__}.{k}", v
+                )
+                # The is safe because we ban overriding `__init__`.
+                instance = component_cls()
+                setattr(self, k, instance)
+                conf[k] = instance
+
+            # Hydrate the sub-component. The configuration we use consists of
+            # all non-scoped keys and any keys scoped to `k`, where the keys
+            # scoped to `k` override the non-scoped keys.
+            non_scoped_conf = {a: b for a, b in conf.items() if "." not in a}
+            k_scoped_conf = {
+                a[len(f"{k}.") :]: b for a, b in conf.items() if a.startswith(f"{k}.")
+            }
+            nested_conf = dict(non_scoped_conf, **k_scoped_conf)
+            getattr(self, k).hydrate(
+                nested_conf,
+                parent=self,
+                name=f"{self.__component_name__}.{k}",
+                interactive=interactive,
+            )
+
+        # Validate all parameters.
+        self.validate_configuration()
```

Comment:
Ah this is true. Originally I had a `validate_param(name, value, annotated_type)` method called on each param but I changed it because I thought it would be nicer for overriding to have a no-args validate method where everything is up to the subclass. But I think the benefit you highlight when running interactively outweighs that, so I'll change it back.

Reply:
I have done the type-checking in `__setattr__` and left in a `validate_configuration` method which by default does nothing but can be overriden.